### PR TITLE
[KAFKA-9644] Non-existent configs in incrementalAlterConfigs APPEND/SUBTRACT

### PR DESCRIPTION
**Pull request by steverod**
Problem
----
The `incrementalAlterConfigs` API supports OpType.APPEND and OpType.SUBTRACT for configuration properties of LIST type. If an APPEND or SUBTRACT OpType is submitted for a config property which currently has no value, then the operation fails with a NullPointerException on the broker side (conveyed as an "unknown server error" to the client).

This is because the alter code does a `getProperty` of the existing configuration value
with no concern as to whether or not the property actually exists.

This change handles the case of existing null properties. 

Testing
-----
This change includes 2 test cases in the unit test that demonstrate the issue for OpType.SUBTRACT and OpType.APPEND. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/server/AdminManager.scala
Comment (steverod) [inline core/src/main/scala/kafka/server/AdminManager.scala]: The filter code is necessary to generate an empty list when there's an empty property.

File: core/src/main/scala/kafka/server/AdminManager.scala
Comment (steverod) [inline core/src/main/scala/kafka/server/AdminManager.scala:622]: An open question is if we subtract everything, should that remove the property or just leave it as an empty string?

Comment (omkreddy) [reply core/src/main/scala/kafka/server/AdminManager.scala:622]: I think, we should remove the config.

File: core/src/test/scala/integration/kafka/api/PlaintextAdminIntegrationTest.scala
Comment (omkreddy) [inline core/src/test/scala/integration/kafka/api/PlaintextAdminIntegrationTest.scala:1660]: Can we add a assert statement below to verify that config got added?

File: core/src/test/scala/integration/kafka/api/PlaintextAdminIntegrationTest.scala
Comment (omkreddy) [inline core/src/test/scala/integration/kafka/api/PlaintextAdminIntegrationTest.scala:1688]: Can we add a assert statement to verify config doesn't exist before and after the alter call?

**Review by omkreddy — COMMENTED** [review]
@steverod Thanks for the PR. Left couple of minor comments.

File: core/src/main/scala/kafka/server/AdminManager.scala
Comment (bob-barrett) [reply core/src/main/scala/kafka/server/AdminManager.scala]: Do we need the `trim` call? We should already get a length 0 string back if the property is empty. Seems unlikely that we'd ever have a list value that includes a whitespace-only string as an element, but who knows.

Comment (steverod) [reply core/src/main/scala/kafka/server/AdminManager.scala]: I'll remove and try.

Comment (steverod) [reply core/src/main/scala/kafka/server/AdminManager.scala]: Removed the trim and it still passes tests. yay.

File: core/src/main/scala/kafka/server/AdminManager.scala
Comment (steverod) [reply core/src/main/scala/kafka/server/AdminManager.scala:622]: After spending several hours arguing with the test, I realized that a "removal" (delete) reverts the property to the default. So if we have a property whose list value is "foo, bar" and we SUBTRACT "foo, bar" ... is the property now set to "" or to <default>? (Especially if default IS, say, "foo")

Comment (omkreddy) [reply core/src/main/scala/kafka/server/AdminManager.scala:622]: I think, we should fallback to broker/topic defaults after removal of any overrides. currently, we should the same for  OpType.DELETE operation. We should also follow the same for OpType. SUBTRACT?

Comment (bob-barrett) [reply core/src/main/scala/kafka/server/AdminManager.scala:622]: I think it makes more sense to end up with an empty list if you SUBTRACT each list element. That's how lists typically work. We have DELETE if the desired behavior is to actually remove the config.

Comment (omkreddy) [reply core/src/main/scala/kafka/server/AdminManager.scala:622]: My main concern is setting empty list to configs like `cleanup.policy` may lead to unwanted errors. 

cc @rajinisivaram

Comment (rajinisivaram) [reply core/src/main/scala/kafka/server/AdminManager.scala:622]: May be worth looking at all the broker and topic configs of type LIST and seeing whether empty list makes sense in most cases. If it doesn't. perhaps throwing an exception may be better than deleting quietly since you may end up with a default value that you didn't intend for the config.

Comment (steverod) [reply core/src/main/scala/kafka/server/AdminManager.scala:622]: Thanks @rajinisivaram ! -- a discussion with others led to the realization that empty string is the wrong answer in the general case, I have updated to use the stringified version of the config default.

File: core/src/test/scala/integration/kafka/api/PlaintextAdminIntegrationTest.scala
Comment (steverod) [reply core/src/test/scala/integration/kafka/api/PlaintextAdminIntegrationTest.scala:1688]: Done.

File: core/src/test/scala/integration/kafka/api/PlaintextAdminIntegrationTest.scala
Comment (steverod) [reply core/src/test/scala/integration/kafka/api/PlaintextAdminIntegrationTest.scala:1660]: Done.

File: core/src/main/scala/kafka/server/AdminManager.scala
Comment (steverod) [reply core/src/main/scala/kafka/server/AdminManager.scala:622]: I have just pushed a version which stays with the "subtract from the list even if the list goes to empty", and will momentarily push a version which does the "config removal/revert to default when subtract yields an empty list". I defer to the Kafka committers on this.

**Comment by steverod** [comment]
@omkreddy @rajinisivaram  thank you for the feedback! I have updated the PR to properly handle defaults and to remove the property entirely (thus reverting to default) when a SUBTRACT removes all members of the list.

File: core/src/main/scala/kafka/server/AdminManager.scala
Comment (omkreddy) [reply core/src/main/scala/kafka/server/AdminManager.scala:622]: I think it makes sense to apply append/subtract on the config default
value rather than "", if the config is previously not set. 

The only issue with config removal/revert to default is we can not set empty list to config value.

Comment (steverod) [reply core/src/main/scala/kafka/server/AdminManager.scala:622]: We could always force users to use the `SET` operation in this case, rather than `SUBTRACT`.
However, on thinking about it, the current code will let you SUBTRACT to an empty list. So in the interest of keeping this a bug fix (not breaking existing code/changing behavior, just preventing the NPE), maybe we should not include my last pushed commit and allow SUBTRACT to generate an empty list. This is rope, but it is existing rope.

Comment (omkreddy) [reply core/src/main/scala/kafka/server/AdminManager.scala:622]: makes sense. we can revert last commit and handle separately in another JIRA.

Can we please update the javadoc with current behaviour of `OpType.SET`, 
`OpType.DELETE`, `OpType.APPEND` and `OpType.SUBTRACT` operations?
https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/admin/Admin.java#L394

File: core/src/main/scala/kafka/server/AdminManager.scala
Comment (omkreddy) [inline core/src/main/scala/kafka/server/AdminManager.scala]: `configKeys(configPropName).defaultValue` can return `null`. we may need to treat  `null` as empty string,

Comment (steverod) [reply core/src/main/scala/kafka/server/AdminManager.scala]: Good point.

**Comment by omkreddy** [comment]
add to whitelist

File: core/src/main/scala/kafka/server/AdminManager.scala
Comment (steverod) [reply core/src/main/scala/kafka/server/AdminManager.scala:622]: FYI, validated in the Javadoc for AlterConfigOp.OpType --

image

**Review by omkreddy — APPROVED** [review]
@steverod Thanks for the updates. LGTM.

**Pull request closed without merging**
